File: backend/app/src/main/kotlin/net/animeta/backend/controller/v3/CreateAccountController.kt
package net.animeta.backend.controller.v3

import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.User
import net.animeta.backend.repository.UserRepository
import net.animeta.backend.service.AuthService
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import java.time.Instant

@RestController
class CreateAccountController(
    private val userRepository: UserRepository,
    private val authService: AuthService
) {
    data class Params(
        val username: String,
        val password1: CharArray,
        val password2: CharArray
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Params

            if (username != other.username) return false
            if (!password1.contentEquals(other.password1)) return false
            if (!password2.contentEquals(other.password2)) return false

            return true
        }

        override fun hashCode(): Int {
            var result = username.hashCode()
            result = 31 * result + password1.contentHashCode()
            result = 31 * result + password2.contentHashCode()
            return result
        }
    }

    data class Result(
        val authResult: AuthenticateController.Result
    )

    @PostMapping("/v3/CreateAccount")
    fun create(@RequestBody params: Params): Result {
        val existingUser = userRepository.findByUsername(params.username)
        if (existingUser != null) {
            throw usernameAlreadyExistError()
        }
        if (
            params.username.isEmpty() ||
            params.username.length > 30 ||
            !params.username.matches("^[A-Za-z0-9_]+$".toRegex()) ||
            params.password1.isEmpty() ||
            params.password2.isEmpty() ||
            !params.password1.contentEquals(params.password2)
        ) {
            throw ApiException("회원가입 실패", HttpStatus.BAD_REQUEST)
        }
        val user = User(username = params.username, date_joined = Instant.now())
        authService.setPassword(user, params.password1)
        try {
            userRepository.save(user)
        } catch (e: DataIntegrityViolationException) {
            throw usernameAlreadyExistError()
        }
        val session = authService.createSession(user, persistent = false)
        return Result(
            authResult = AuthenticateController.Result(
                sessionKey = session.sessionKey,
                expiryMs = session.expiry?.toMillis()
            )
        )
    }

    private fun usernameAlreadyExistError() =
        ApiException("이미 사용 중인 아이디입니다. 다른 아이디로 가입하세요.", HttpStatus.BAD_REQUEST)
}

File: backend-ts/src/controllers/create_account.controller.ts
import { Body, Controller, HttpStatus, Post } from "@nestjs/common";
import { User } from "src/entities/user.entity";
import { AuthService } from "src/services/auth.service";
import { ApiException } from "./exceptions";
import { Result as AuthResult } from "./authenticate.controller";
import { QueryFailedError, Repository } from "typeorm";
import { InjectRepository } from "@nestjs/typeorm";

type Params = {
  username: string;
  password1: string;
  password2: string;
};

type Result = {
  authResult: AuthResult;
};

@Controller('/api/v4/CreateAccount')
export class CreateAccountController {
  constructor(
    @InjectRepository(User) private userRepository: Repository<User>,
    private authService: AuthService,
  ) {}

  @Post()
  async handle(@Body() params: Params): Promise<Result> {
    const existingUser = await this.userRepository.findOne({ where: {username: params.username} })
    if (existingUser)
      throw usernameAlreadyExistError()
    if (
      params.username === '' ||
      params.username.length > 30 ||
      !/^[A-Za-z0-9_]+$/.test(params.username) ||
      params.password1 === '' ||
      params.password2 === '' ||
      params.password1 !== params.password2
    ) {
      throw new ApiException("회원가입 실패", HttpStatus.BAD_REQUEST)
    }
    const user = new User()
    user.username = params.username
    user.first_name = ''
    user.last_name = ''
    user.email = ''
    user.is_staff = false
    user.is_active = true
    user.is_superuser = false
    user.last_login = new Date()
    user.date_joined = new Date()
    await this.authService.setPassword(user, params.password1)
    try {
      await this.userRepository.save(user)
    } catch (e) {
      // https://www.postgresql.org/docs/10/errcodes-appendix.html
      if (e instanceof QueryFailedError && e.driverError?.code === '23505') {
        throw usernameAlreadyExistError()
      } else {
        throw e
      }
    }
    const session = this.authService.createSession(user, false)
    return {
      authResult: {
        sessionKey: session.sessionKey,
        expiryMs: session.expiry?.total({unit: 'millisecond'}) ?? null,
      }
    }
  }
}

function usernameAlreadyExistError() {
  return new ApiException("이미 사용 중인 아이디입니다. 다른 아이디로 가입하세요.", HttpStatus.BAD_REQUEST)
}
